pulumitest/cacheStackState.go

```diff
@@ -0,0 +1,78 @@
+package pulumitest
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/pulumi/providertest/pulumitest/opttest"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+)
+
+// CacheStackState caches the stack state from executing the `run` func. Subsequent calls will load the stack state from the cache.
+// If the cache does not exist, it will run the `run` func and cache the resulting stack state.
+// The `run` function is always executed in an isolated temp directory so it won't affect the original test.
+// The state exported from the `run` function (or cached from the previous run) is imported into the original test's stack.
+func (pulumiTest *PulumiTest) CacheStackState(cachePath string, run func(test *PulumiTest), opts ...opttest.Option) *PulumiTest {
+	pulumiTest.T().Helper()
+	if cachePath == "" {
+		pulumiTest.T().Fatal("cachePath is required")
+	}
+
+	stackExport, err := TryReadStackExport(cachePath)
+	if err != nil {
+		pulumiTest.T().Fatalf("failed to read stack export: %v", err)
+	}
+	if stackExport != nil {
+		pulumiTest.T().Logf("load stack state from cache %s", cachePath)
+	} else {
+		pulumiTest.T().Logf("no stack state cache found at %s", cachePath)
+		cacheRecording := pulumiTest.CopyToTempDir()
+		run(cacheRecording)
+		cacheRecording.T().Logf("writing stack state to %s", cachePath)
+		exportedStack := cacheRecording.ExportStack()
+		err = WriteStackExport(cachePath, &exportedStack)
+		if err != nil {
+			cacheRecording.T().Fatalf("failed to write snapshot to %s: %v", cachePath, err)
+		}
+		stackExport = &exportedStack
+	}
+	pulumiTest.ImportStack(*stackExport)
+	return pulumiTest
+}
+
+// WriteStackExport writes the stack export to the given path creating any directories needed.
+func WriteStackExport(path string, snapshot *apitype.UntypedDeployment) error {
+	if snapshot == nil {
+		return fmt.Errorf("stack export must not be nil")
+	}
+	dir := filepath.Dir(path)
```

Comment:
Would it make sense here, for being extra safe, to check there is no such file yet? If I read the code correctly, it would be silently overwritten.